# Notebook sobre listas

Este notebook va a ser sobre ejercicios de listas recuperados del libro: *Introducción a la programacion con python 3* de, Andrés Mazal Varó, Isabel Garcia Luengo, Pedro Garcia Sevilla.

Creado por: Rodrigo Salazar @sv_roo

## Ejercicio 215

Haz un programa que almacen en una variable *a* a la lista obtenida mediante *list(range(1,4))* y, a continuacón, la modifique para que cada componente sea igual al cuadrado del componente original. El programa mostrará la lista resultante por pantalla.

In [8]:
# Declaracion de la lista
a = list(range(1,4))
i = 0

# Mostramos en pantalla el contenido antes del calculo
print("Lista antes de modificar\n", a)

# Realizamos el calculo del cuadrado de cada item
while(i < 3):
    a[i] = a[i] ** 2
    i +=1

"""
Se muestra solucion con un ciclo for por si no se quiere usar un ciclo while
   for i in range(0, len(a)):
    a[i] = a[i] ** 2
"""
    
# Mostramos en pantalla el calculo del cuadrado de cada item de la lista
print("Lista despues de modificar\n", a)

Lista antes de modificar
 [1, 2, 3]
Lista despues de modificar
 [1, 4, 9]


## Ejercicio 216

Haz un programa que almacene en una variable *a* la lista obtenida con *list(range(1,n))*, donde *n* es un entero que se pide al usuario, y modifique dicha lista para que cada componente sea igual al cuadrado del componente original. El programa mostrará la lista resultante por pantalla.

In [9]:
# Declaracion de variables 
n = int(input("Dame el limite de una lista:"))

# Declaramos la lista con el rango del usuario 
a = list(range(1, n))

# Mostramos la lista antes del calculo
print("Lista antes de modificar:\n", a)

# Procedemos con el calculo del cuadrado de cada elemento con ciclo for
for i in range(0, len(a)):
    a[i] = a[i] * a[i]
    
# Mostramos la lista después de calcular
print("Lista modificada:\n", a)

Lista antes de modificar:
 [1, 2, 3, 4]
Lista modificada:
 [1, 4, 9, 16]


## Problema 217 

Haz un programa que, dada una lista *a* cualquiera, sustituya cualquier elelemtno negativo por cero.

In [14]:
# Declaracion de variables
a = []
limite = int(input("Dame el limite de la lista que vas a llenar:\n"))
i = 0

# Procedemos a pedir los datos de la lista
while (i <= limite):
    n = int(input("Dame un número:\n"))
    a.append(n)
    i += 1
    
    
# MOstramos lista antes de modificar
print("Lista antes de modificar:\n", a)
# Procedemos a validar cada elemento para ver si es menor que cero
for i in range(0,len(a)):
    if a[i] < 0:
        a[i] = 0
    
# Mostramos lista después de modificar
print("Lista modificada:\n", a)

Lista antes de modificar:
 [-10, -1, 0, 1, 2, 3]
Lista modificada:
 [0, 0, 0, 1, 2, 3]


## Ejemplo

Vamos a esciribir un programa que construya una lista con todos los números primos entre 1 y n. Como no sabemos a priori cuántos hay, construiremos una lsita vaciía e iremos añadiendo números primos conforme los vayamos encontrando.

In [17]:
# Declaracion de variables
n = int(input("Introduce el valor máximo:\n"))
primos = []

# Procedimiento para buscar primos
for numero in range(2, n + 1):
    # Determinamos si un número es primo
    creo_que_es_primo = True
    for divisor in range(2, numero):
        if numero % divisor == 0:
            creo_que_es_primo = False
            break
    if creo_que_es_primo:
        primos.append(numero)
        
# Mostramos en pantalla la lista 
print(primos)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201]


## Ejercicio 220

Diseña un programa que constuya una lista con los *n* primeros números primos (ojo: no los primos entre 1 y *n*, sino los *n* primeros números primos) ¿Necesitas usar append? ¿Puedes reservar en primer lugar una lista con *n* celdas nulas y asignarle a cada una de ellas uno de los números primos?

In [36]:
# Declaracion de lista vacia y limite de primos
n = int(input("Ingresa el limite de numero primos"))
primos = []
i = 0

# Iniciamos el proceso
for numero in range(2, 10000):
    # Determinamos si un número es primo
    creo_que_es_primo = True
    for divisor in range(2, numero):
        if numero % divisor == 0:
            creo_que_es_primo = False
            break
    if creo_que_es_primo:
        i += 1
        if i <= n:
            primos.append(numero)
                    
# Imprimimos la lista 
print(primos)

[2]


## Ejercicio 221

Diseña un programa que lea una lsita de 10 enteros, pero asegurándose de que todos los números introducidos por el usuario son positivos. Cuando un número sea negativo, lo indicaremos con un mensaje y permitiremos al usuario repetir el intento cuantas veces sea preciso.

In [40]:
# Declaracion de varialbles
numeros = []
i = 0

# Proceso para guardar numeros y validar sean mayores de 0
while (i < 10):
    numero = int(input("Dame un número:\n"))
    if(numero > 0):
        numeros.append(numero)
    else:
        while(numero < 0):
            numero = int(input("Dame un número:\n"))
        numeros.append(numero)
    i += 1

# Mostramos la lista
print(numeros)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


## Ejercicio 222

Diseña un programa que lea una cadena y muestre por pantalla una lista con todas sus palabras en minúsculas. La lista no debe contener palabras repetidas.

In [50]:
# Declaracion de variables
palabra = input("Ingresa una cadena:\n")

# Usamos split para separar las palbras
separado = palabra.split(" ")

# Convertimos la lista a conjunto para eliminar repetidos
separada = set(separado)

# Convertimos a lista
separada = list(separada)

#Mostramos el resultado
print(separada)

['palabras', 'una', 'otras', 'formada', 'con', 'frase', 'otra']


## Ejemplo 

Eliminar elementos negativos de una lista

In [51]:
# Declaracion de variables
a = [1, 2, -1, -4, 5, -2]
i = 0

# Proceso para eliminar
while i < len(a):
    if a[i] < 0:
        del a[i]
    else:
        i += 1

# Imprimimos la lista
print(a)

[1, 2, 5]


## Ejercicio 224

Diseña un programa que elimine de una lista todos los elementos de índice par y muestre por pantalla el resultado

*(Ejemplo: si trabaja con la lista [1, 2, 1, 5, 0, 3], está pasara a ser [2, 5, 3])*

In [104]:
# Declaracion de variables
a = [1, 2, 1, 5, 0, 3]
i = 0

# Inicio del proceso de eliminacion
#Mostramos lista antes de eliminar
print("Antes de eliminar:\n", a)
while i < len(a):
    if i % 2 == 0:
        del a[i]
        i += 1
    else:
        if (i + 1) % 2 == 0:
            del a[i]
        i += 1
            
# Imprimimos la lista 
print("Despues de eliminar\n", a)

Antes de eliminar:
 [1, 2, 1, 5, 0, 3]
Despues de eliminar
 [2, 5, 3]


Codigo Linea por linea para entender como funciona el codigo anterior donde muestra cada eliminacion que se hace y cada valor de las variables involucradas.

(La lista se puede modificar de tamaño y va a seguir funcionando sin problemas)

In [107]:
# Declaracion de variables
a = [1, 2, 1, 5, 0, 3]
i = 0

# Inicio del proceso de eliminacion
print("Lista original:", a)

while i < len(a):
    print("\n\n\tCiclo número:", i + 1)
    print("i:", i)
    print("len: ", len(a))
    if i % 2 == 0:
        print("Lista con i par:", a)
        del a[i]
        print("Despues de eliminar desde el if:", a)
        i += 1
    else:
        if (i + 1) % 2 == 0:
            print("Lista con i impar:", a)
            del a[i]
            print("Despues de eliminar desde el false:", a)
        i += 1
            
# Imprimimos la lista 
print(a)


Lista original: [1, 2, 1, 5, 0, 3]


	Ciclo número: 1
i: 0
len:  6
Lista con i par: [1, 2, 1, 5, 0, 3]
Despues de eliminar desde el if: [2, 1, 5, 0, 3]


	Ciclo número: 2
i: 1
len:  5
Lista con i impar: [2, 1, 5, 0, 3]
Despues de eliminar desde el false: [2, 5, 0, 3]


	Ciclo número: 3
i: 2
len:  4
Lista con i par: [2, 5, 0, 3]
Despues de eliminar desde el if: [2, 5, 3]
[2, 5, 3]


## Ejercicio 225

Diseña un programa que elimine de una lista todos los elementos de valor par y muestre en pantalla el resultado.

*(Ejemplo: si trabaja con la lista [1, -2, 1, -5, 0, 3], esta pasará a ser [1, 1, -5, 3])*

In [4]:
# Declaracion de variables
lista = [1, -2, 1, -5, 0, 3]

#Inicio del proceso

#Mostramos la lista
print(lista)
for i in lista:
    if lista[i] % 2 == 0:
        del lista[i]
        i += 1
    else:
        if lista[i + 1] % 2 == 0:
            del lista[i]
            i += 1
        
print(lista)

[1, -2, 1, -5, 0, 3]
[1, 1, -5, 3]


## Ejemplo
### Pertenencia de un elemento a una lista

Diseñemos un programa que, dados un elemento y una lista, nos diga si el elemento pertenece o no a la lista mostrando en pantalla el mensaje *Pertenece* o *No pertenece* en función del resultado

In [5]:
#Declaracion de variables
elemento = 5
lista = [1, 4, 5, 1, 3, 8]

#Proceso
pertenece = False
for i in lista:
    if elemento == i:
        pertenece = True
        break
    
if pertenece:
    print("Pertenece")
else:
    print("No pertenece")

Pertenece


### 227 ¿Por qué este otro programa es erróneo?

In [7]:
# Declaracion de variables
elemento = 5
lista = [1, 4, 5, 1, 3, 8]

for i in lista:
    if elemento == i:
        pertenece = True
    else:
        pertenece = False
        
if pertenece:
    print("Pertenece")
else:
    print("No pertenece")
    
    """Esta mal ya que al usar el else en este caso en particular el ultimo elemento de la lista no es el 5 es un 8 por lo que toma el valor de false
    sin embargo si esta el elemento en la lista
    """

No pertenece


La pregunta de si un elemento pertenece o no a una lista es tan frecuente que Python nos
proporciona un operador predefinido que hace eso mismo. El operador es binario y se denota con
la palabra in (que en inglés significa «en» o «pertenece a»). El operador in recibe un elemento
por su parte izquierda y una lista por su parte derecha y devuelve cierto o falso. Un programa
que necesita determinar si un elemento pertenece o no a una lista y actuar en consecuencia
puede hacerlo así:

In [8]:
conjunto = [1, 2, 3]
elemento = int(input("Dame un número:\n"))

if not elemento in conjunto:
    conjunto.append(elemento)

print(conjunto)

[1, 2, 3, 5]


O, equivalente

In [9]:
conjunto = [1, 2, 3]
elemento = int(input("Dame un número:\n"))

if elemento not in conjunto:
    conjunto.append(elemento)

print(conjunto)

[1, 2, 3, 7]


## Metodo de ordenamiento burbuja


In [11]:
# Declaracion de variable
lista = [2, 26, 4, 3, 1]

# Mostramos la lista antes del proceso
print(lista)

# Inicio del proceso
for i in range(1, len(lista)):
    for j in range(0, len(lista) - 1):
        if lista[j] > lista [j + 1]:
            elemento = lista[j]
            lista[j] = lista [j + 1]
            lista[j+1] = elemento
            
# Mostramos la lista ordenada
print(lista)
    

[2, 26, 4, 3, 1]
[1, 2, 3, 4, 26]


In [12]:
#Comrpobacion del programa para ver que hace cada paso
# Declaracion de variable
lista = [2, 26, 4, 3, 1]

# Mostramos la lista antes del proceso
print(lista)

# Inicio del proceso
for i in range(1, len(lista)):
    print("Pasada: ", i)
    for j in range(0, len(lista) - 1):
        print("Comparacion de lista {0} y lista {1}". format(j, j+1))
        if lista[j] > lista [j + 1]:
            elemento = lista[j]
            lista[j] = lista [j + 1]
            lista[j+1] = elemento
            print("Se intercambian")
        print("Estado actual de la lista", lista)
            
# Mostramos la lista ordenada
print(lista)
    

[2, 26, 4, 3, 1]
Pasada:  1
Comparacion de lista 0 y lista 1
Estado actual de la lista [2, 26, 4, 3, 1]
Comparacion de lista 1 y lista 2
Se intercambian
Estado actual de la lista [2, 4, 26, 3, 1]
Comparacion de lista 2 y lista 3
Se intercambian
Estado actual de la lista [2, 4, 3, 26, 1]
Comparacion de lista 3 y lista 4
Se intercambian
Estado actual de la lista [2, 4, 3, 1, 26]
Pasada:  2
Comparacion de lista 0 y lista 1
Estado actual de la lista [2, 4, 3, 1, 26]
Comparacion de lista 1 y lista 2
Se intercambian
Estado actual de la lista [2, 3, 4, 1, 26]
Comparacion de lista 2 y lista 3
Se intercambian
Estado actual de la lista [2, 3, 1, 4, 26]
Comparacion de lista 3 y lista 4
Estado actual de la lista [2, 3, 1, 4, 26]
Pasada:  3
Comparacion de lista 0 y lista 1
Estado actual de la lista [2, 3, 1, 4, 26]
Comparacion de lista 1 y lista 2
Se intercambian
Estado actual de la lista [2, 1, 3, 4, 26]
Comparacion de lista 2 y lista 3
Estado actual de la lista [2, 1, 3, 4, 26]
Comparacion de lis

## 230 ¿Qué pasará si sustituimos la primera línea de burbuja.py por esta otra?

lista = ["Pepe", "Juan", "Maria", "Ana", "Luis", "Pedro"]

In [13]:
# Declaracion de variable
lista = ["Pepe", "Juan", "Maria", "Ana", "Luis", "Pedro"]

# Mostramos la lista antes del proceso
print(lista)

# Inicio del proceso
for i in range(1, len(lista)):
    for j in range(0, len(lista) - 1):
        if lista[j] > lista [j + 1]:
            elemento = lista[j]
            lista[j] = lista [j + 1]
            lista[j+1] = elemento
            
# Mostramos la lista ordenada
print(lista)
    

['Pepe', 'Juan', 'Maria', 'Ana', 'Luis', 'Pedro']
['Ana', 'Juan', 'Luis', 'Maria', 'Pedro', 'Pepe']


# 5.4 Matrices

Las matrices son disposiciones bidimensionales de valores. En notación matemática, una matriz se denota encerrando entre paréntesis los valores, que se disponen en filas y columnas.

He aquí una matriz M:

$$
\left(
\begin{array}{ll}
1 & 2 & 3 \\
2 & 12 & 6 \\
1 & 0 & -3 \\
0 & -1 & 0
\end{array}
\right)
$$

Esta matriz tiene 4 filas y 3 columnas, lo cual abreviamos diciendo que es una matriz de dimensión 4x3

Las listas permiten representar series de datos en una sola dimensión. Con una lista de
números no se puede representar directamente una matriz, pero sí con una lista de listas de
números.

*M = [[1, 2, 3], [2, 12, 6], [1, 0, -3], [0, -1, 0]]*

### 235:
Una matriz nula es aquella que solo contiene cero. Construye una matriz nula de 5 filas y 5 columnas

In [22]:
# Declaracion de variables
mtz_nula = [[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

# Mostramos la matriz en pantalla
for i in range(0, 5):
    print(mtz_nula[i])

[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]


### 236:
Una amtriz identidad es aquella cuyos elementos en la diagonal principal, es decir, accesibles con una expresión de la forma M[i][j] valen uno y el resto valen cero. Construye una matriz identidad de 4 filas y 4 columnas

In [23]:
# Declaracion de variables
mtz_nula = [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]

# Mostramos la matriz
for i in range(0, 4):
    print(mtz_nula[i])

[1, 0, 0, 0]
[0, 1, 0, 0]
[0, 0, 1, 0]
[0, 0, 0, 1]


In [21]:
# ¿Que resulta de ejecutar este programa?

# Declaracion de variables
M = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
print(M[-1][0])
print(M[-1][1])
print("-")
for i in range(0, 3):
    print(M[i])
print("-")
for i in range(0, 3):
    for j in range(0, 3):
        print(M[i][j])

0
0
-
[1, 0, 0]
[0, 1, 0]
[0, 0, 1]
-
1
0
0
0
1
0
0
0
1


In [24]:
# ¿Que resulta de ejecutar este programa?
M = [[1, 0, 1], [0, 1, 0], [0, 0, 1]]
s = 0.0
for i in range(0,3):
    for j in range(0, 3):
        s += M[i][j]
        
print(s / 9)

0.4444444444444444


Creacion de matrices para no crear uno a uno la matriz nula

In [34]:
M = []
for i in range (3):
    a = [0] * 6
    M.append(a)

print(M)

[[0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0]]


In [36]:
M = []
for i in range (3):
    M.append([0] * 6)

M[0][1] = 1
print(M)

[[0, 1, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0]]


### Ejercicio 239

Crea la siguiente matriz utilizando la técnica del bucle descrita anteriormente
$$
\left(
\begin{array}{ll}
1 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1 \\
\end{array}
\right)
$$

In [38]:
# Declaracion de variables
M = []

# Inicio del primer ciclo para crear la matriz de ceros
for i in range(4):
    M.append([0] * 4)

# Ciclo para llenar la diagonal principal de unos
for i in range(4):
    M[i][i] = 1
    
# Mostramos la matriz por pantalla
for i in M:
    print(i)

[1, 0, 0, 0]
[0, 1, 0, 0]
[0, 0, 1, 0]
[0, 0, 0, 1]


### Ejercicio 240

Haz un programa que pida un entero positivo n y almacene en una variable M la matriz identidad n x n (la que tiene unos en la diagonal principal y ceros en el resto de celdas)

In [62]:
# Declaracion de variables
M = []
num = int(input("Dame un numero:\n"))

# Inicio del primer ciclo para crear la matriz de ceros
for i in range(num):
    M.append([0] * num)

# Ciclo para llenar la diagonal principal de unos
for i in range(num):
    M[i][i] = 1
    
# Mostramos la matriz por pantalla
for i in M:
    print(i)

[1, 0]
[0, 1]


### 5.4.2 Lectura de matrices

Si deseamos leer una matriz de tamaño determindado, podemos crear un matriz nula como hemos visto en el apartado anterior y a continuación rellenar cada uno de sus componentes:

In [67]:
# Pedimos la dimension de la matriz
m = int(input('Dime el numero de filas:'))
n = int(input('Dame el numero de columnas'))

# Creamos una matriz nula
M = []
for i in range(m):
    M.append( [0] * n)

# y leemos lsu contenido de teclado
for i in range(m):
    for j in range(n):
        M[i][j] = float(input('Dame el componente {0},{1}'.format(i, j)))
        
# Mostramos la matriz
for i in M:
    print(i)

[1.0, 2.0]
[3.0, 4.0]


## 5.4.4 Operaciones con matrices

Desarrollaremos ahora algunos programas que nos ayuden a efectuar operaciones con matrices como la suma o el producto

In [74]:
# Pedimos la dimension de las matrices
m = int(input('Dime el numero de filas:'))
n = int(input('Dime el numero de columnas:'))

# Creamos dos matrices nulas
A = []
for i in range(m):
    A.append( [0] * n)

B = []
for i in range(m):
    B.append( [0] * n )
    
# y tleemos sus contenidos de teclado
print('Lectura de la matriz A')
for i in range(m):
    for j in range(n):
        A[i][j] = float(input('Componente {0}{1}'.format(i, j)))

print('Lectura de la matriz b')
for i in range(m):
    for j in range(n):
        B[i][j] = float(input('Componente {0}{1}'.format(i, j)))
        
# Construimos otra matriz nula para albergar el resultado
C = []
for i in range(m):
    C.append([0] * n)

# Empezamos el calculo de la suma
for i in range(m):
    for j in range(n):
        C[i][j] = A[i][j] + B[i][j]
        
# y mostramos el resultado por pantalla
print('Suma:')
for i in C:
    print(i)

ValueError: invalid literal for int() with base 10: ''

## Ejercicio 241 

Diseña un programa que lea dos matrices y calcule la diferencia entre la primera y la segunda

In [70]:
# Pedimos la dimension de las matrices
m = int(input('Dime el numero de filas:'))
n = int(input('Dime el numero de columnas:'))

# Creamos dos matrices nulas
A = []
for i in range(m):
    A.append( [0] * n)

B = []
for i in range(m):
    B.append( [0] * n )
    
# y leemos sus contenidos de teclado
print('Lectura de la matriz A')
for i in range(m):
    for j in range(n):
        A[i][j] = float(input('Componente {0}{1}'.format(i, j)))

print('Lectura de la matriz b')
for i in range(m):
    for j in range(n):
        B[i][j] = float(input('Componente {0}{1}'.format(i, j)))
        
# Construimos otra matriz nula para albergar el resultado
C = []
for i in range(m):
    C.append([0] * n)

# Empezamos el calculo de la suma
for i in range(m):
    for j in range(n):
        C[i][j] = A[i][j] - B[i][j]
        
# y mostramos el resultado por pantalla
print('Resta:')
for i in C:
    print(i)

Lectura de la matriz A
Lectura de la matriz b
Resta:
[-4.0, -4.0]
[-4.0, -4.0]


### Ejercicio 242

Diseña un programa que lea una matriz y un numero y devuelva una nueva matriz la que resulta de multiplicar la matriz por el numero. (El producto de un numero por una matriz es la matriz que resulta de multiplicar cada elemento por dicho numero)

In [71]:
# Pedimos la dimension de las matrices
m = int(input('Dime el numero de filas:'))
n = int(input('Dime el numero de columnas:'))
numero = float(input('Dame el numero por el cual multiplicar:\n'))

# Creamos dos matrices nulas
A = []
for i in range(m):
    A.append( [0] * n)

# y leemos sus contenidos de teclado
print('Lectura de la matriz A')
for i in range(m):
    for j in range(n):
        A[i][j] = float(input('Componente {0}{1}'.format(i, j)))
        

# Creamos matriz nula para almacenar los resultados
C = []
for i in range(m):
    C.append([0] * n)
    
# Creamos la multiplicacion de la matriz por n
for i in range(m):
    for j in range(n):
        C[i][j] = A[i][j] * numero
        
for i in C:
    print(i)

Lectura de la matriz A
[1.0, 2.0]
[3.0, 4.0]


## Multiplicacion de matrices


In [73]:
# Pedimos la dimension de la primera matriz y el numero de columnas de la segunda
p = int(input('Dime el numero de filas de A:\n'))
q = int(input('Dime el numero de columnas A (y filas de B:)\n'))
r = int(input('Dime el numero de columnas de B"\n'))

# Creamos dos matrices nulas
A = []
for i in range(p):
    A.append([0] * q)
    
B = []
for i in range(q):
    B.append([0] * r)
    
# y leeremos contenido del teclado
print('lectura de la matriz A')
for i in range(p):
    for j in range(q):
        A[i][j] = float(input('Componente {0}, {1}'.format(i, j)))

print('lectura de la matriz B')
for i in range(q):
    for j in range(r):
        B[i][j] = float(input('Componente {0}, {1}'.format(i, j)))
        
# Creamos una matriz nula para el resultado
C = []
for i in range(p):
    C.append([0] * r)
    
# Efectuamos el calculo del producto
for i in range(p):
    for j in range(r):
        for k in range(q):
            C[i][j] += A[i][k] * B[k][j]
            
            
# Mostramos resultados en pantalla
for i in C:
    print(i)

lectura de la matriz A
lectura de la matriz B
[19.0, 22.0]
[43.0, 50.0]


## Ejercicio 243

La transpuesta de una matriz A de dimensión mxn es una matriz A<sup>T</sup> de dimension nxm tal que A<sup>T</sup> <sub>i ° j</sub>  = A <sub>j * i</sub>

$$A=
\left(
\begin{array}{ll}
1 & 2 & 3 \\
2 & 12 & 6 \\
1 & 0 & -3 \\
10 & -1 & 0 
\end{array}
\right)
$$

entonces:

$$A^T =
\left(
\begin{array}{ll}
1 & 2 & 1 & 10 \\
2 & 12 & 0 & -1 \\
3 & 6 & -3 & 0
\end{array}
\right)
$$

Diseña un programa que lea una matriz y muestre su transpuesta

In [78]:
# Declaracion de variables
n = int(input('Ingresa el numero de columnas:\n'))
m = int(input('Ingresa el numero de filas:\n'))

#Incializamos la matriz nula
A = []
for i in range(m):
    A.append([0] * n)

# Pedimos datos desde el teclado
for i in range(m):
    for j in range(n):
        A[i][j] = float(input('Dame el elemento {0}, {1}:\n'.format(i, j)))

# Inicializamos la matriz donde vamos a almacenar el resultado
B = []
for i in range(n):
    B.append([0] * m)
    
# Realizamos la transpuesta
for i in range(n):
    for j in range(m):
        B[i][j] = A[j][i]
    
#Mostramos la matris normal
print('Matriz A:')
for i in A:
    print(i)
       
# Mostramos la Inversa de la matriz
print('Inversa de la matriz A:')
for i in B:
    print(i)

Matriz A:
[1.0, 2.0]
[3.0, 4.0]
[5.0, 6.0]
[7.0, 8.0]
Inversa de la matriz A:
[1.0, 3.0, 5.0, 7.0]
[2.0, 4.0, 6.0, 8.0]


### Ejercicio 244

Diseña un programa tal que lea una matriz A de dimension mxn y muestre un vector v de talla n tal que 

vj = m
i=1
Ai,j ,


para j entre 1 y n

In [6]:
# Declaramos variables
p = int(input('Dime el numero de filas de A:\n'))
q = int(input('Dime el numero de columnas A:\n'))
aux = 0
vector = []

# Inicializamos la matriz con ceros
A = []
for i in range(p):
    A.append([0] * q)
    
#se piden datos por teclado
for i in range(p):
    for j in range(q):
        A[i][j] = float(input('Dame el elemento {0}, {1}:\n'.format(i, j)))
        
# Procedemos a llenar el vector
for i in range(p):
    for j in range(q):
        aux = A[i][j] + aux
    vector.append([aux])
    aux = 0
        
# Imprimimos las dos matrices para comprobar
print('\tMatriz generada:')
for i in A:
    print(i)
    
print('\tVector:')
for i in vector:
    print(i)

	Matriz generada:
[10.0, 15.0, 20.0, 0.0]
[2.0, 15.0, 14.0, 20.0]
[23.0, 25.0, 22.0, 56.0]
[17.0, 15.0, 10.0, 1.0]
	Vector:
[45.0]
[51.0]
[126.0]
[43.0]


### Ejercicio 245 

Diseña un programa que lea una matriz A de dimensión m x n y muestre un vextor v de talla min(n,m) tal que sea la sumatoria de j =1 hasta i y k = 1 hasta i de A <sub>j.k</sub> para i entre 1 u min(n,m)

In [8]:
# Declaramos variables
p = int(input('Dime el numero de filas de A:\n'))
q = int(input('Dime el numero de columnas A:\n'))
aux = 0
vector = []

# Inicializamos la matriz vacia
A = []
for i in range(p):
    A.append([0] * q)

# Empezamos a llenar la matriz con datos del teclado
for i in range(p):
    for j in range(q):
        A[i][j] = float(input('Dame el elemento {0}, {1}:\n'.format(i, j)))

# Empezamos a realizar la sumatoria
for i in range(p):
    for j in range(q):
        aux = A[i][j] + aux
    vector.append([aux])
    
# Imprimimos matriz y vector y el min(m, n)
print(min(p,q))
print("\tMatriz")
for i in A:
    print(A)

print("\tVector:")
for i in vector:
    print(i)
    

2
	Matriz
[[1.0, 2.0], [3.0, 4.0]]
[[1.0, 2.0], [3.0, 4.0]]
	Vector:
[3.0]
[10.0]


### Ejercicio 246 

Diseañ un program que, dada una matriz determine si la suma de los elementos de cualquiera de sus filas es igual a la suma de los elementos de cualquiera de sus columnas

In [15]:
# Declaramos variables
f = int(input("Ingresa el numero de filas:\n"))
c = int(input("Ingresa el numero de columnas:\n"))
aux = 0
suma_fila = []
suma_columna = []
coincide = False

# Inicializamos la matriz con ceros
A = []
for i in range(f):
    A.append([0] * c)
    
# Pedimos datos desde el teclado
for i in range(p):
    for j in range(q):
        A[i][j] = float(input('Dame el elemento {0}, {1}:\n'.format(i, j)))
        
# Realizamos las sumas por fila y se almacenan en una lsita
for i in range(p):
    for j in range(q):
        aux = A[i][j] + aux
    suma_fila.append([aux])
    aux = 0

# Realizamos las sumas por columna y almacenarlos en una lista
aux = 0
for i in range(p):
    for j in range(q):
        aux = A[j][i] + aux
    suma_columna.append([aux])
    aux = 0
        
# Comparamos los resultados y si uno coincide lo mostramos
for i in range(1, len(suma_columna)):
    for j in range(0, len(suma_columna) - 1):
        if suma_columna[j] == suma_fila[j + 1]:
            coincide = True

if coincide:
    print("Hay sumas iguales entre filas y columnas")
else:
    print("Ninguna suma coincide")
    

Hay sumas iguales entre filas y columnas


### 5.4.5 El juego de la vida

El juego de la vida es un juego sin jugadores. Se trata de colocar una serie de fichas en un tablero y dejar que evolucionen siguiendo unas reglas extremadamente simples. Lo curioso es que esas reglas dan origen a una gran complejidad que hace apasionante la mera observación de la evolución de las fichas en el tablero (hay gustos para todo).
En el juego original se utiliza un tablero (una matriz) con infinitas filas y columnas. Como disponer de una matriz de dimensión infinita en un programa es imposible, supondremos que presenta dimensión m x n, donde m y n son valores escogidos por nosotros. Cada celda del tablero contiene una célula que puede estar viva o muerta. Representaremos las células vivas con su casilla de color negro y las células muertas con la celda en blanco. Cada casilla del tablero cuenta con ocho celdas vecinas. El mundo del juego de la vida está gobernado por un reloj que marca una serie de pulsos con los que mueren y nacen células. Cuándo nace y cuándo muere una célula solo depende de cuántas células vecinas están vivas. He aquí las reglas:

- Regla del nacimiento. Una célula muerta resucita si tiene exactamente tres vecinos vivos. En
estas figuras te señalamos celdas muertas que pasan a estar vivas con el siguiente pulso
- Regla de la supervivencia. Una celda viva permanece viva si tiene dos o tres vecinos. Aquí te
señalamos células que ahora están vivas y permanecerán así tras el siguiente pulso
- Regla de la superpoblación. Una célula muere o permanece muerta si tiene cuatro o más
vecinos. Estas figuras muestran células que ahora están vivas o muertas y estarán muertas
tras el siguiente pulso
- Regla del aislamiento. Una célula muere o permanece muerta si tiene menos de dos vecinos. En estas figuras te señalamos células que ahora están vivas o muertas y estarán muertas tras el siguiente pulso

Vamos a hacer un programa que muestre la evolución del juego de la vida durante una serie de pulsos de reloj. Empezaremos con un prototipo que nos muestra la evolución del tablero en el terminal.

In [22]:
filas = 10
columnas = 10

tablero = []
for i in range(filas):
    tablero.append([False] * columnas)

tablero[4][5] = True
tablero[5][5] = True
tablero[6][5] = True

for y in range(filas):
    for x in range(columnas):
        if tablero[y][x]:
            print("*", end="")
        else:
            print(".", end="")
    print()
    
pulsos = 6
for t in range(pulsos):
    
    # Preparar un nuevo tablero
    nuevo = []
    for i in range(filas):
        nuevo.append([False] * columnas)
        
    # Actualizar el tablero
    for y in range(filas):
        for x in range(columnas):
            # Calcular el numero de vecinos de la celda que estamos visitando
            n = 0
            if y > 0 and x > 0 and tablero[y-1][x-1]:
                n += 1
            if x > 0 and tablero[y][x-1]:
                n +=1
            if y < filas - 1 and x > 0 and tablero[y+1][x-1]:
                n += 1
            if y > 0 and tablero[y-1][x]:
                n += 1
            if y < filas - 1 and tablero[y+1][x]:
                n +=1
            if y > 0 and x < columnas - 1 and tablero[y-1][x+1]:
                n += 1
            if x < columnas - 1 and tablero[y][x+1]:
                n +=1
            if y < filas - 1 and x < columnas - 1 and tablero[y+1][x+1]:
                n += 1
            # aplicar las reglas
            if tablero[y][x] and (n == 2 or n == 3): #Supervivencia
                nuevo[y][x] = True
            elif not tablero[y][x] and n == 3: # Nacimiento
                nuevo[y][x] = True
            else: # Superpoblacion y aislamiento
                nuevo[y][x] = False       
                                             
    # Actualizar el teclado
    tablero = nuevo
    
    # Representar el tablero
    print("Pulso", t+1)
    for y in range(filas):
        for x in range(columnas):
            if tablero[y][x]:
                print("*", end="")
            else:
                print(".", end="")
        print()

..........
..........
..........
..........
.....*....
.....*....
.....*....
..........
..........
..........
Pulso 1
..........
..........
..........
..........
..........
....***...
..........
..........
..........
..........
Pulso 2
..........
..........
..........
..........
.....*....
.....*....
.....*....
..........
..........
..........
Pulso 3
..........
..........
..........
..........
..........
....***...
..........
..........
..........
..........
Pulso 4
..........
..........
..........
..........
.....*....
.....*....
.....*....
..........
..........
..........
Pulso 5
..........
..........
..........
..........
..........
....***...
..........
..........
..........
..........
Pulso 6
..........
..........
..........
..........
.....*....
.....*....
.....*....
..........
..........
..........
